nrf802154_sniffer.py

```diff
@@ -0,0 +1,445 @@
+#
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# 3. Neither the name of Nordic Semiconductor ASA nor the names of its
+#    contributors may be used to endorse or promote products derived from this
+#    software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+import os
+
+is_standalone = __name__ == '__main__'
+
+if sys.version[0] == '2':
+    # py2 support
+    import Queue as Queue
+else:
+    import queue as Queue
+
+if __name__ == '__main__':
+    sys.path.insert(0, os.getcwd())
+
+
+import re
+import select
+import signal
+import struct
+import threading
+import time
+import logging
+from argparse import ArgumentParser
+from binascii import a2b_hex
+from serial import Serial, serialutil
+from serial.tools.list_ports import comports
+
+
+class Nrf802154Sniffer(object):
+
+    # Various options for pcap files: http://www.tcpdump.org/linktypes.html
+    #DLT='user'
+    DLT='802.15.4'
+    DLT_NO = 147 if DLT == 'user' else 230
+
+    # helper for wireshark arg parsing
+    CTRL_ARG_CHANNEL = 0
+
+    # pattern for packets being printed over serial
+    RCV_REGEX = 'received:\s+([0-9a-fA-F]+)\s+power:\s+(-?\d+)\s+lqi:\s+(\d+)\s+time:\s+(-?\d+)'
+
+    def __init__(self):
+        self.serial_queue = Queue.Queue()
+        self.running = threading.Event()
+        self.setup_done = threading.Event()
+        self.setup_done.clear()
+        self.logger = logging.getLogger(__name__)
+        self.dev = None
+        self.channel = None
+
+    def stop_sig_handler(self, *args, **kwargs):
+        """
+        Function responsible for stopping the sniffer firmware and closing all threads.
+        """
+        # Let's wait with closing afer we're sure that the sniffer started. Protects us
+        # from very short tests (NOTE: the serial_reader has a delayed started)
+        while self.running.is_set() and not self.setup_done.is_set():
+            time.sleep(1)
+
+        if self.running.is_set():
+            self.serial_queue.put(b'')
+            self.serial_queue.put(b'sleep')
```

Comment:
Is this needed? I found this problematic when quickly stopping and then starting the sniffer from python script. (it resulted in `did not reply properly to setup commands. Is it flashed properly?`)

Reply:
Yes, it is needed. You need to stop receiver before you stop sniffing.
If you don't stop receiver before you stop sniffer, receiver keeps putting frames to buffers. Then you have old frames in UART buffers, USB buffers, PHY buffers. When you restart sniffing (without power cycling of the sniffer device), buffers are flushed and you get list of old frames before actual sniffing starts.

Reply:
BTW, there is a delay before sniffer is started in order to overcome this problem. Is this delay too short?